File: backend/app/schemas/practice.py
from uuid import UUID

from pydantic import BaseModel, Field, ValidationInfo, field_validator

from app.core.sanitize import sanitize_text
from app.models.practice_log import Grade
from app.schemas.llm import SentenceReview


class PracticeCardRead(BaseModel):
    card_id: UUID
    word: str
    context_sentence: str | None = None
    previous_sentence: str | None = None


class DailyPracticeResponse(BaseModel):
    session_id: UUID
    cards: list[PracticeCardRead]


class SentenceSubmit(BaseModel):
    card_id: UUID
    user_sentence: str = Field(..., max_length=1000)
    revealed_translation: bool = False

    @field_validator("user_sentence", mode="before")
    @classmethod
    def validate_user_sentence(cls, v: str) -> str:
        if not isinstance(v, str):
            return v
        v = sanitize_text(v)
        if not v:
            raise ValueError("Sentence cannot be empty")
        return v


class PracticeSubmitRequest(BaseModel):
    session_id: UUID
    sentences: list[SentenceSubmit]

    @field_validator("sentences")
    @classmethod
    def check_sentences_length(
        cls, v: list[SentenceSubmit], info: ValidationInfo
    ) -> list[SentenceSubmit]:
        if len(v) < 1:
            raise ValueError("At least one sentence is required")
        return v


class PracticeLogReview(BaseModel):
    id: UUID
    card_word: str
    user_sentence: str
    grade: Grade | None = None
    llm_feedback: SentenceReview | None = None


class PracticeSessionReviewResponse(BaseModel):
    session_id: UUID
    logs: list[PracticeLogReview]

File: frontend/src/api/practice.ts
import { apiClient } from "./client";

export interface PracticeCardRead {
    card_id: string;
    word: string;
    context_sentence: string | null;
    previous_sentence: string | null;
}

export interface DailyPracticeResponse {
    session_id: string;
    cards: PracticeCardRead[];
}

export interface SentenceSubmit {
    card_id: string;
    user_sentence: string;
    revealed_translation: boolean;
}

export interface PracticeSubmitRequest {
    session_id: string;
    sentences: SentenceSubmit[];
}

export const practiceApi = {
    getDailyPractice: async (): Promise<DailyPracticeResponse> => {
        return apiClient("/practice/daily");
    },

    submitPractice: async (data: PracticeSubmitRequest): Promise<void> => {
        return apiClient("/practice/submit", {
            method: "POST",
            body: JSON.stringify(data),
        });
    },
};
